evm/p2p/discovery.py

```diff
@@ -0,0 +1,320 @@
+    pass
+
+
+class WrongMAC(DefectiveMessage):
+    pass
+
+
+class PacketExpired(DefectiveMessage):
+    pass
+
+
+class Command():
+    def __init__(self, name, id, elem_count):
+        self.name = name
+        self.id = id
+        # Number of required top-level list elements for this cmd.
+        # Elements beyond this length must be trimmed.
+        self.elem_count = elem_count
+
+    def __repr__(self):
+        return 'Command(%s:%d)' % (self.name, self.id)
+
+
+CMD_PING = Command("ping", 1, 4)
+CMD_PONG = Command("pong", 2, 3)
+CMD_FIND_NODE = Command("find_node", 3, 2)
+CMD_NEIGHBOURS = Command("neighbours", 4, 2)
+CMD_ID_MAP = dict((cmd.id, cmd) for cmd in [CMD_PING, CMD_PONG, CMD_FIND_NODE, CMD_NEIGHBOURS])
+
+
+class DiscoveryProtocol(asyncio.DatagramProtocol):
+    """A Kademlia-like protocol to discover RLPx nodes."""
+    logger = logging.getLogger("evm.p2p.discovery.DiscoveryProtocol")
+    transport = None
+    _max_neighbours_per_packet_cache = None
+
+    def __init__(self, privkey, address, bootstrap_nodes):
+        self.privkey = privkey
+        self.pubkey = private_key_to_public_key(self.privkey)
+        self.address = address
+        self.bootstrap_nodes = bootstrap_nodes
+        self.this_node = kademlia.Node(self.pubkey, address)
+        self.kademlia = kademlia.KademliaProtocol(self.this_node, wire=self)
+
+    def _get_handler(self, cmd):
+        if cmd == CMD_PING:
+            return self.recv_ping
+        elif cmd == CMD_PONG:
+            return self.recv_pong
+        elif cmd == CMD_FIND_NODE:
+            return self.recv_find_node
+        elif cmd == CMD_NEIGHBOURS:
+            return self.recv_neighbours
+        else:
+            raise ValueError("Unknwon command: {}".format(cmd))
+
+    def _get_max_neighbours_per_packet(self):
+        # As defined in https://github.com/ethereum/devp2p/blob/master/rlpx.md, the max size of a
+        # datagram must be 1280 bytes, so when sending neighbours packets we must include up to
+        # _max_neighbours_per_packet and if there's more than that split them across multiple
+        # packets.
+        if self._max_neighbours_per_packet_cache is not None:
+            return self._max_neighbours_per_packet_cache
+        # Use an IPv6 address here as we're interested in the size of the biggest possible node
+        # representation.
+        addr = kademlia.Address('::1', 30303, 30303)
+        node_data = addr.to_endpoint() + [b'\x00' * (kademlia.k_pubkey_size // 8)]
+        neighbours = [node_data]
+        expiration = rlp.sedes.big_endian_int.serialize(int(time.time() + EXPIRATION))
+        payload = rlp.encode([neighbours] + [expiration])
+        while HEAD_SIZE + len(payload) <= 1280:
+            neighbours.append(node_data)
+            payload = rlp.encode([neighbours] + [expiration])
+        self._max_neighbours_per_packet_cache = len(neighbours) - 1
+        return self._max_neighbours_per_packet_cache
```

Comment:
Other than the code at the head of the function which returns the cached value, this function body looks like it could be a stand-alone function.

Reply:
Right, done